internal/allocator/allocator.go

```diff
@@ -194,113 +187,118 @@ func (a *Allocator) Unassign(svc string) bool {
-		delete(a.sharingKeyForIP, al.ip.String())
-	}
-	a.poolIPsInUse[al.pool][al.ip.String()]--
-	if a.poolIPsInUse[al.pool][al.ip.String()] == 0 {
-		// Explicitly delete unused IPs from the pool, so that len()
-		// is an accurate count of IPs in use.
-		delete(a.poolIPsInUse[al.pool], al.ip.String())
 	}
 	stats.poolActive.WithLabelValues(al.pool).Set(float64(len(a.poolIPsInUse[al.pool])))
 	return true
 }
 
-func cidrIsIPv6(cidr *net.IPNet) bool {
-	return cidr.IP.To4() == nil
-}
-func ipIsIPv6(ip net.IP) bool {
-	return ip.To4() == nil
-}
-
 // AllocateFromPool assigns an available IP from pool to service.
-func (a *Allocator) AllocateFromPool(svc string, isIPv6 bool, poolName string, ports []Port, sharingKey, backendKey string) (net.IP, error) {
+func (a *Allocator) AllocateFromPool(svc string, serviceIPFamily ipfamily.Family, poolName string, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
 	if alloc := a.allocated[svc]; alloc != nil {
 		// Handle the case where the svc has already been assigned an IP but from the wrong family.
-		// This "should-not-happen" since the "ipFamily" is an immutable field in services.
-		if isIPv6 != ipIsIPv6(alloc.ip) {
-			return nil, fmt.Errorf("IP for wrong family assigned %s", alloc.ip.String())
+		// This "should-not-happen" since the "serviceIPFamily" is an immutable field in services.
+		allocIPsFamily, err := ipfamily.ForAddressesIPs(alloc.ips)
+		if err != nil {
+			return nil, err
+		}
+		if allocIPsFamily != serviceIPFamily {
+			return nil, fmt.Errorf("IP for wrong family assigned alloc %s service family %s", allocIPsFamily, serviceIPFamily)
 		}
-		if err := a.Assign(svc, alloc.ip, ports, sharingKey, backendKey); err != nil {
+		if err := a.Assign(svc, alloc.ips, ports, sharingKey, backendKey); err != nil {
 			return nil, err
 		}
-		return alloc.ip, nil
+		return alloc.ips, nil
 	}
 
 	pool := a.pools[poolName]
 	if pool == nil {
 		return nil, fmt.Errorf("unknown pool %q", poolName)
 	}
+	sk := &key{
+		sharing: sharingKey,
+		backend: backendKey,
+	}
+	ips := []net.IP{}
+	ipfamilySel := make(map[ipfamily.Family]bool)
+
+	switch serviceIPFamily {
+	case ipfamily.DualStack:
+		ipfamilySel[ipfamily.IPv4], ipfamilySel[ipfamily.IPv6] = true, true
+	default:
+		ipfamilySel[serviceIPFamily] = true
+	}
 
 	for _, cidr := range pool.CIDR {
-		if cidrIsIPv6(cidr) != isIPv6 {
+		cidrIPFamily := ipfamily.ForCIDR(cidr)
+		if _, ok := ipfamilySel[cidrIPFamily]; !ok {
 			// Not the right ip-family
 			continue
 		}
-		c := ipaddr.NewCursor([]ipaddr.Prefix{*ipaddr.NewPrefix(cidr)})
-		for pos := c.First(); pos != nil; pos = c.Next() {
-			ip := pos.IP
-			if pool.AvoidBuggyIPs && ipConfusesBuggyFirmwares(ip) {
-				continue
-			}
-			// Somewhat inefficiently brute-force by invoking the
-			// IP-specific allocator.
-			if err := a.Assign(svc, ip, ports, sharingKey, backendKey); err == nil {
-				return ip, nil
-			}
+		ip := a.getIPFromCIDR(cidr, pool.AvoidBuggyIPs, svc, ports, sk)
```

Comment:
nit: I'd have `getIPFromCIDR` accept sharingkey and backendKey and construct the sk inside (or use sk also in allocate, but that's a bit more complex)